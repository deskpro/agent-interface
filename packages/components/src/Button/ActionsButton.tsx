import * as React from "react";

import Button, { ButtonProps } from "./Button";
import Portal from "../common/Portal/Portal";
import useWindowSize from "../hooks/useWindowSize";
import useOutsideClick from "../hooks/useOutsideClick";
import useMenu from "../hooks/useMenu";
import { MenuProps } from "../elements/Menu/Menu";

export type ActionsButtonProps = ButtonProps & {
  renderMenu: (menuProps: Partial<MenuProps>) => React.ReactElement<any>;
};

function useMenuPosition(buttonRef, menuRef, isVisible) {
  const windowSize = useWindowSize();
  const [menuStyles, setStyles] = React.useState({ top: 0, left: 0 });

  React.useEffect(
    () => {
      if (!isVisible) {
        return;
      }
      const styles = { ...menuStyles };
      if (buttonRef.current instanceof HTMLButtonElement) {
        styles.left = buttonRef.current.offsetLeft;
        styles.top =
          buttonRef.current.offsetTop + buttonRef.current.offsetHeight + 3;

        if (menuRef.current instanceof HTMLUListElement) {
          const { width: windowWidth, height: windowHeight } = windowSize;
          if (
            windowWidth &&
            styles.left + menuRef.current.offsetWidth > windowWidth
          ) {
            styles.left -=
              menuRef.current.offsetWidth - buttonRef.current.offsetWidth;
          }
          if (
            windowHeight &&
            styles.top + menuRef.current.offsetHeight > windowHeight
          ) {
            styles.top -=
              menuRef.current.offsetHeight + buttonRef.current.offsetHeight;
          }
        }
      }

      setStyles(styles);
    },
    [buttonRef, menuRef, isVisible]
  );

  return menuStyles;
}

const ActionsButton = ({ renderMenu, ...props }) => {
  const { isVisible: menuIsVisible, hideMenu, toggleMenu } = useMenu();

  const buttonRef = React.useRef<HTMLButtonElement>(null);
  const menuRef = React.useRef<HTMLUListElement>(null);
  useOutsideClick(buttonRef, hideMenu);
  const menuStyles = useMenuPosition(buttonRef, menuRef, menuIsVisible);

  return (
    <>
      <Button
        {...props}
        onClick={toggleMenu}
        variant="actions"
        buttonRef={buttonRef}
      />
      <Portal>
        {menuIsVisible &&
          renderMenu({
            menuRef,
            style: menuStyles,
            onMenuClose: hideMenu
          })}
      </Portal>
    </>
  );
};

export default ActionsButton;
