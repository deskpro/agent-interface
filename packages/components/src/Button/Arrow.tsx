import * as React from "react";
import classNames from "classnames";

import "@deskpro/agent-interface-style/dist/dp-Arrow.css";

export type ArrowProps = {
  isActive?: boolean;
  onClick?: (e: React.SyntheticEvent) => void;
};

export default React.forwardRef<HTMLSpanElement, ArrowProps>(
  ({ isActive, onClick }, ref) => (
    <span
      className={classNames("dp-Icon dp-Arrow", { "is-active": isActive })}
      ref={ref}
      role="button"
      tabIndex={-1}
      onClick={onClick}
      onKeyPress={e => e.key === "Enter" && onClick && onClick(e)}
    />
  )
);
