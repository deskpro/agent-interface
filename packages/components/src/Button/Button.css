@import "../fonts.css";
@import "../vars.css";

.dp-Button {
  margin-bottom: 10px;
}

.dp-Button .Button--large {
  font-size: var(--font-size-md);
  max-width: 140px;
  padding: 10px 1rem;
  text-align: center;
  width: 100%;
}

.dp-Button.Button--in-active {
  background: var(--gray-in-active);
  color: var(--white);
}

.dp-Button.Button--active {
  background: var(--blue-active);
  color: var(--white);
}

.dp-Button.Button--rollover {
  background: var(--blue-rollover);
  color: var(--white);
}

.dp-Button.Button--small {
  max-width: 54px;
}

.Button--small.Button--in-active {
  background: var(--white);
  color: var(--gray-600);
  opacity: 0.5;
}

.Button--in-active {
  cursor: not-allowed;
}

.dp-ButtonDescription {
  align-items: center;
  display: flex;
  justify-content: space-between;
}

.dp-ButtonState {
  color: var(--gray);
  font-size: var(--font-size-xs);
  margin-bottom: 10px;
}

.dp-ButtonItem .dp-ActionButton {
  border: none;
  margin: 0 0 10px;
  max-width: 86px;
}

.dp-ActionButton.Button--in-active {
  opacity: 0.5;
}

.dp-ActionButton.Button--active {
  background: #eff0f0;
}

.dp-ActionButton.Button--rollover {
  background: #d3d6d7;
}

.dp-ExpandButton.Button--in-active {
  background: var(--gray-in-active);
  cursor: pointer;
}

.dp-ExpandButton.dp-ExpandButton-small {
  border-radius: var(--border-radius-small);
  max-width: 16px;
}

.dp-Button .dp-IconPlusBtn {
  background-size: 10px 10px;
  height: 10px;
  width: 10px;
}

.dp-Button.dp-ButtonPlus {
  align-items: center;
  background: var(--gray-50);
  display: flex;
  justify-content: space-between;
  padding-left: 5px;
  padding-right: 5px;
}

.dp-Button.dp-LabelRollover {
  background: #eff0f0;
}

.dp-BorderPink {
  border-color: var(--pink);
}

.dp-Pink {
  color: var(--pink);
}

.dp--danger {
  color: var(--danger);
}

.dp-Border--danger {
  border-color: var(--danger);
}

.dp-Border--brand-primary {
  border-color: var(--brand-primary);
}

/* main buttons style */

.dp-Button {
  background: #ccc;
  background: var(--white);
  border: 1px solid var(--gray-200);
  border-radius: var(--border-radius);
  color: var(--text-dark);
  cursor: pointer;
  display: inline-block;
  font-size: var(--font-Title-S);
  line-height: 100%;
  padding: var(--button-padding) 1rem;

  &:hover {
    background: var(--gray-100);
  }

  &.Button--white {
    background: var(--white);
    border: 1px solid var(--white);
    color: var(--text-dark);

    &:hover {
      background: var(--gray-100);
    }
  }

  &.Button--transparent {
    background: transparent;
    border: 1px solid #6e8293;
    color: #6e8293;

    &:hover {
      background: var(--gray-100);
    }
  }

  &.Button--text {
    background: transparent;
    border: 1px solid transparent;
    color: var(--text-dark);
    text-decoration: underline;

    &:hover {
      background: var(--gray-100);
    }
  }

  &.Button--black {
    background: var(--black);
    border: 1px solid var(--black);
    color: var(--text-light);

    &:hover {
      background: var(--gray-500);
    }
  }

  &.Button--primary {
    background: var(--color-primary);
    border: 1px solid var(--color-primary);
    color: var(--text-light);

    &:hover {
      background: color-mod(var(--color-primary) blackness(20%));
    }
  }

  &.Button--secondary {
    background: var(--color-secondary);
    border: 1px solid var(--color-secondary);
    color: var(--text-light);

    &:hover {
      background: color-mod(var(--color-secondary) blackness(20%));
    }
  }

  &.Button--success {
    background: var(--color-success);
    border: 1px solid var(--color-success);
    color: var(--text-light);

    &:hover {
      background: color-mod(var(--color-success) blackness(20%));
    }
  }

  &.Button--info {
    background: var(--color-info);
    border: 1px solid var(--color-info);
    color: var(--text-light);

    &:hover {
      background: color-mod(var(--color-info) blackness(20%));
    }
  }

  &.Button--warning {
    background: var(--color-warning);
    border: 1px solid var(--color-warning);
    color: var(--text-light);
    color: var(--text-dark);

    &:hover {
      background: color-mod(var(--color-warning) blackness(20%));
    }
  }

  &.Button--danger {
    background: var(--color-danger);
    border: 1px solid var(--color-danger);
    color: var(--text-light);

    &:hover {
      background: color-mod(var(--color-danger) blackness(20%));
    }
  }

  &.Button--rounded {
    border-radius: 100px;
  }

  &[disabled] {
    cursor: not-allowed;
    opacity: 0.55;
  }

  &.Button--small {
    border-radius: calc(var(--border-radius) * 0.7);
    font-size: var(--font-Text-N);

    /* font-size: calc(var(--font-size-base) * 0.7); */
    padding: calc(0.7rem * 0.7) calc(1rem * 0.7);
  }

  &.Button--medium {
    border-radius: calc(var(--border-radius) * 1.2);
    font-size: calc(var(--font-size-base) * 1.2);
    padding: calc(0.7rem * 1.2) calc(1rem * 1.2);
  }

  &.Button--large {
    border-radius: calc(var(--border-radius) * 1.4);
    font-size: var(--font-Title-S);

    /* font-size: calc(var(--font-size-base) * 1.4); */
    padding: calc(0.7rem * 1.4) calc(1rem * 1.4);
  }
}

.is-semibold {
  font-weight: var(--is-semibold);
}
