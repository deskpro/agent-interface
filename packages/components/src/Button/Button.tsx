import * as React from "react";
import classNames from "classnames";
import "@deskpro/agent-interface-style/dist/elements/dp-buttons.css";

export interface ButtonProps {
  size?: "default" | "small";
  variant?:
    | "outlined"
    | "contained"
    | "label"
    | "more"
    | "actions"
    | "link"
    | "round"
    | "square";
  color?: DPColor;
  className?: string;
  buttonRef?: React.Ref<HTMLButtonElement>;
  onClick(e: React.MouseEvent<HTMLElement>): void;
}

const Button: React.FC<ButtonProps> = ({
  onClick: handleClick,
  children,
  size = "default",
  variant = "contained",
  color = "primary",
  className,
  buttonRef,
  ...props
}) => (
  <button
    type="button"
    className={classNames(
      "dp-Button",
      {
        [`Button--${size}`]: !!size && size !== "default",
        [`Button--${variant}`]: !!variant && variant !== "contained",
        [`Button--${color}`]: !!color
      },
      className
    )}
    ref={buttonRef}
    onClick={handleClick}
    {...props}
  >
    {variant === "more" ? <span className="dp-dots">...</span> : children}
  </button>
);

Button.defaultProps = {
  size: "default",
  variant: "contained"
};

export default React.forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => <Button {...props} buttonRef={ref} />
);
