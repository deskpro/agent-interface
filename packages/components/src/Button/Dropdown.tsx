import * as React from "react";
import { Manager, Reference, Popper } from "react-popper";
import classNames from "classnames";

import "@deskpro/agent-interface-style/dist/dp-Arrow.css";
import { MenuProps } from "../elements/Menu/Menu";
import useOutsideClick from "../hooks/useOutsideClick";

export type DropdownProps = {
  className?: string;
  renderMenu: (menuProps: Partial<MenuProps>) => React.ReactElement<any>;
};

const Dropdown: React.FC<DropdownProps> = ({
  className,
  children,
  renderMenu
}) => {
  const [isOpen, setOpenState] = React.useState(false);
  const dropdownRef = React.useRef<HTMLSpanElement>(null);
  useOutsideClick(dropdownRef, () => {
    setOpenState(false);
  });

  return (
    <span className="dp-Dropdown-container" ref={dropdownRef}>
      <Manager>
        <Reference>
          {({ ref }) => (
            <span
              className={classNames("dp-Dropdown", className)}
              ref={ref}
              role="button"
              tabIndex={-1}
              onClick={() => setOpenState(!isOpen)}
              onKeyPress={e => {
                if (e.key === "Enter") setOpenState(!isOpen);
              }}
            >
              {children}
              <span className="dp-Arrow" />
            </span>
          )}
        </Reference>
        {isOpen && (
          <Popper
            placement="bottom-start"
            modifiers={{
              preventOverflow: {
                enabled: true,
                escapeWithReference: true,
                boundariesElement: "viewport"
              },
              flip: {
                enabled: true,
                flipVariationsByContent: true
              }
            }}
          >
            {({ ref, style }) =>
              renderMenu({
                menuRef: ref,
                style,
                onMenuClose: () => setOpenState(false)
              })
            }
          </Popper>
        )}
      </Manager>
    </span>
  );
};

export default Dropdown;
