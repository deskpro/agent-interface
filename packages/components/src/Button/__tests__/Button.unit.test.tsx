import * as React from "react";
import { render, fireEvent } from "react-testing-library";

import Button from "../Button";

describe("<Button />", () => {
  it("Simple snapshot", () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <Button onClick={handleClick}>Simple Button</Button>
    );
    expect(getByText("Simple Button")).toMatchSnapshot();
  });
  it("More variant snapshot", () => {
    const handleClick = jest.fn();
    const { getByTestId } = render(
      <Button data-testid="more-button" variant="more" onClick={handleClick}>
        More Button
      </Button>
    );
    expect(getByTestId("more-button")).toMatchSnapshot();
  });
  it("should render attributes", () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <Button onClick={handleClick}>Simple Button</Button>
    );

    expect(getByText("Simple Button")).toHaveAttribute("type", "button");
  });
  it("should call onClick fn on click", () => {
    const handleClick = jest.fn();
    const { getByText } = render(
      <Button onClick={handleClick}>Simple Button</Button>
    );

    fireEvent(
      getByText("Simple Button"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true
      })
    );
    expect(handleClick.mock.calls.length).toBe(1);
  });
});
