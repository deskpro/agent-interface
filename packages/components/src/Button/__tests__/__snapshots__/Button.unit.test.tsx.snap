// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> More variant snapshot 1`] = `
<button
  class="dp-Button Button--more Button--primary"
  data-testid="more-button"
  type="button"
>
  <span
    class="dp-dots"
  >
    ...
  </span>
</button>
`;

exports[`<Button /> Simple snapshot 1`] = `
<button
  class="dp-Button Button--primary"
  type="button"
>
  Simple Button
</button>
`;
